interfaces/prompting/requestrules/requestrules.go

```diff
@@ -307,24 +340,156 @@ func (rdb *RuleDB) addRuleToRulesList(rule *Rule) error {
+// returns an error, and the rule DB is left unchanged.
+//
+// The caller must ensure that the database lock is held for writing.
+func (rdb *RuleDB) addOrMergeRule(rule *Rule, save bool) (addedOrMergedRule *Rule, merged bool, err error) {
+	// Check if rule with identical path pattern exists.
+	existingRule, exists, err := rdb.lookupRuleByPathPattern(rule.User, rule.Snap, rule.Interface, rule.Constraints)
+	if err != nil {
+		// Database was left inconsistent, should not occur
+		return nil, false, err
+	}
+	if !exists {
+		return rule, false, rdb.addNewRule(rule, save)
+	}
+
+	newPermissions := make(prompting.RulePermissionMap)
+	// Add any non-expired permissions from the existing rule. Each might later
+	// be overridden by a permission entry in the new rule, if the latter has a
+	// broader lifespan (and doesn't otherwise conflict).
+	for existingPerm, existingEntry := range existingRule.Constraints.Permissions {
+		if existingEntry.Expired(rule.Timestamp) {
+			continue
+		}
+		newPermissions[existingPerm] = existingEntry
+	}
+	// Check whether the new rule has outcomes/lifespans which conflict with the
+	// existing rule, otherwise add them to the new set of permissions.
+	var conflicts []prompting_errors.RuleConflict
+	for perm, entry := range rule.Constraints.Permissions {
+		existingEntry, exists := newPermissions[perm]
+		if !exists {
+			newPermissions[perm] = entry
+			continue
+		}
+		if entry.Outcome != existingEntry.Outcome {
+			// New entry outcome conflicts with outcome of existing entry
+			conflicts = append(conflicts, prompting_errors.RuleConflict{
+				Permission:    perm,
+				Variant:       rule.Constraints.PathPattern.String(), // XXX: we're mis-using the full path pattern in place of the variant
+				ConflictingID: existingRule.ID.String(),
+			})
+			continue
+		}
+		// Both new and existing rule have the same permission with the same
+		// outcome, so preserve whichever entry has the greater lifespan.
+		// Since newPermissions[perm] already has the existing entry, only
+		// override it if the new rule has a greater lifespan.
+		if prompting.FirstLifespanGreater(entry.Lifespan, entry.Expiration, existingEntry.Lifespan, existingEntry.Expiration) {
+			newPermissions[perm] = entry
+		}
+	}
+	// If there were any conflicts with the existing rule with identical path
+	// pattern, return error.
+	if len(conflicts) > 0 {
+		return nil, false, &prompting_errors.RuleConflictError{
+			Conflicts: conflicts,
+		}
+	}
+
+	// Create new rule based on the contents of the existing rule, but copy the
+	// timestamp from the new rule.
+	newRule := new(Rule)
+	*newRule = *existingRule
+	newRule.Timestamp = rule.Timestamp
+	// Copy constraints as well, since copying the rule just copied the pointer
+	newRule.Constraints = new(prompting.RuleConstraints)
+	*(newRule.Constraints) = *(existingRule.Constraints)
+	// Now set the permissions of the copied constraints to newPermissions
+	newRule.Constraints.Permissions = newPermissions
```

Comment:
this is less confusing IMO:
```suggestion
	newRule.Constraints = &prompting.RuleConstraints{
		PathPattern: rule.Constraints.PathPattern,
		Permissions: newPermissions,
	}
```

Reply:
Agreed. The rationale was that if the fields of `Rule.Constraints` changed, we'd automatically copy all of them no matter what they are, and then we'd set `Permissions` manually, but what you propose is much clearer.